LoopKit/HealthKitSampleStore.swift

```diff
@@ -157,98 +157,180 @@ public class HealthKitSampleStore {
             // If we are now looking farther back, then reset the query
             if oldValue > observationStart {
                 log.default("observationStart changed: creating HK query")
-                createQuery()
+                createQuery(from: "observartionStart")
+            }
+        }
+    }
+
+    private enum QueryAnchorState: Equatable {
+        case uninitialized
+        case initializationComplete(HKQueryAnchor?)
+
+        var anchor: HKQueryAnchor? {
+            switch self {
+            case .uninitialized:
+                return nil
+            case.initializationComplete(let anchor):
+                return anchor
             }
         }
     }
 
     /// The last-retreived anchor from an anchored object query
     internal var queryAnchor: HKQueryAnchor? {
         get {
-            return lockedQueryAnchor.value
+            if case .initializationComplete(let anchor) = queryState.anchorState {
+                return anchor
+            } else {
+                return nil
+            }
         }
         set {
-            var changed: Bool = false
-            lockedQueryAnchor.mutate { (anchor) in
-                if anchor != newValue {
-                    anchor = newValue
-                    changed = true
-                }
-            }
-            if changed {
-                queryAnchorDidChange()
+            mutateQueryState { state in
+                state.anchorState = .initializationComplete(newValue)
+                state.authorizationDetermined = !authorizationRequired
             }
         }
     }
-    private let lockedQueryAnchor: Locked<HKQueryAnchor?>
 
-    func queryAnchorDidChange() {
-        // Subclasses can override
+    private struct QueryState: Equatable {
+        var anchorState: QueryAnchorState
+        var authorizationDetermined: Bool
+
+        // ObserverQuery will fail if HKAuthorizationStatus is undetermined
+        var readyToQuery: Bool {
+            return anchorState != .uninitialized && authorizationDetermined
+        }
+    }
```

Comment:
[general comment] are these changes covered by unit tests? If not, please add some.

Reply:
Yes, this code is covered in a number of tests that exercise store initialization with health kit authorization.